src/v/storage/tests/compaction_e2e_test.cc

```diff
@@ -224,6 +301,419 @@ TEST_P(CompactionFixtureParamTest, TestDedupeOnePass) {
+    }
+}
+
+INSTANTIATE_TEST_SUITE_P(
+  NumSegments,
+  CompactionFixtureTombstonesParamTest,
+  ::testing::Values(1, 10, 100));
+
+struct TombstonesRandomParamArgs {
+    static TombstonesRandomParamArgs create() {
+        static constexpr size_t max_segments = 100;
+        static constexpr size_t max_records = 1000;
+        static constexpr size_t max_cardinality = max_records;
+        static constexpr size_t max_batches_per_segment = 5;
+        return TombstonesRandomParamArgs{
+          .num_segments = random_generators::get_int(size_t{1}, max_segments),
+          .total_records = random_generators::get_int(size_t{1}, max_records),
+          .cardinality = random_generators::get_int(size_t{1}, max_cardinality),
+          .batches_per_segment = random_generators::get_int(
+            size_t{1}, max_batches_per_segment)};
+    }
+
+    size_t num_segments;
+    size_t total_records;
+    size_t cardinality;
+    size_t batches_per_segment;
+};
+
+class CompactionFixtureTombstonesRandomParamTest
+  : public CompactionFixtureTest
+  , public ::testing::WithParamInterface<
+      std::tuple<TombstonesRandomParamArgs, bool>> {};
+
+TEST_P(
+  CompactionFixtureTombstonesRandomParamTest,
+  TestTombstonesRandomDistribution) {
+    const auto [data_args, wait_for_retention_ms] = GetParam();
+    const auto [num_segments, total_records, cardinality, batches_per_segment]
+      = data_args;
+
+    const auto num_batches = num_segments * batches_per_segment;
+    // May not divide evenly.
+    const auto records_per_batch = std::max(
+      size_t{1}, total_records / num_batches);
+
+    map_t latest_kv_map;
+    generate_random_assorted_data(
+      num_segments,
+      cardinality,
+      batches_per_segment,
+      records_per_batch,
+      true,
+      &latest_kv_map)
+      .get();
+
+    auto num_tombstones_produced = std::accumulate(
+      latest_kv_map.begin(),
+      latest_kv_map.end(),
+      0,
+      [](size_t acc, const auto& p) { return acc + size_t{p.second.empty()}; });
+
+    auto num_records_produced = latest_kv_map.size() - num_tombstones_produced;
+
+    std::optional<std::chrono::milliseconds> tombstone_retention_ms
+      = wait_for_retention_ms ? 1000ms
+                              : std::optional<std::chrono::milliseconds>{};
+
+    // Helper function to perform sliding window compaction.
+    auto do_sliding_window_compact = [tombstone_retention_ms](auto& log) {
+        // Compact, allowing the map to grow as large as we need.
+        ss::abort_source never_abort;
+        storage::compaction_config cfg(
+          log->segments().back()->offsets().get_base_offset(),
+          tombstone_retention_ms,
+          ss::default_priority_class(),
+          never_abort,
+          std::nullopt,
+          std::nullopt,
+          nullptr,
+          nullptr);
+        auto& disk_log = dynamic_cast<storage::disk_log_impl&>(*log);
+        disk_log.sliding_window_compact(cfg).get();
```

Comment:
Can we also add some testing for where we drop the compaction's max cardinality, so we cover cases where it takes multiple passes over the log to fully compact? As is, IIUC windowed compaction will use a large enough map to compact every segment